libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/RequireMapboxNavigation.kt

```diff
@@ -0,0 +1,139 @@
+@file:JvmName("RequireMapboxNavigation")
+
+package com.mapbox.navigation.core.lifecycle
+
+import androidx.lifecycle.DefaultLifecycleObserver
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleOwner
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.MapboxNavigation
+import kotlin.properties.ReadOnlyProperty
+import kotlin.reflect.KProperty
+
+/**
+ * Extension function to make it simple to create the [RequireMapboxNavigationDelegate]. Below are
+ * a couple examples of how you may use the delegate.
+ *
+ * Default can be used when [MapboxNavigationApp] is setup elsewhere.
+ * ```
+ * val mapboxNavigation by requireMapboxNavigation()
+ * ```
+ *
+ * Initialize the [MapboxNavigationApp] when you are ready to use it
+ * ```
+ * val mapboxNavigation by requireMapboxNavigation {
+ *   MapboxNavigationApp.setup(..)
+ * }
+ * ```
+ *
+ * Register subscriptions and setup MapboxNavigationApp
+ * ```
+ * private val mapboxNavigation by requireMapboxNavigation(
+ *   onResumedObserver = object : MapboxNavigationObserver {
+ *     override fun onAttached(mapboxNavigation: MapboxNavigation) {
+ *       mapboxNavigation.registerLocationObserver(locationObserver)
+ *       mapboxNavigation.registerRoutesObserver(routesObserver)
+ *     }
+ *     override fun onDetached(mapboxNavigation: MapboxNavigation) {
+ *       mapboxNavigation.unregisterLocationObserver(locationObserver)
+ *       mapboxNavigation.unregisterRoutesObserver(routesObserver)
+ *     }
+ *   }
+ * ) {
+ *   MapboxNavigationApp.setup(
+ *       NavigationOptions.Builder(this)
+ *           .accessToken(accessToken)
+ *           .build()
+ *     )
+ * }
+ * ```
+ *
+ * @see [RequireMapboxNavigationDelegate] for more details.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
```

Comment:
There are a couple odd behaviors found while testing. 

 - Because it lazy loads, `onInitialize` is not called by default the class needs to reference the property
 - You cannot reference the property inside `Activity.onCreate`, you must use `Lifecycle.onCreate` callback. This creates an odd initialize behavior

Considering this is a "require", i've made this delegate _not lazy_. A similar "lazyMapboxNavigation" could be made if the lazy approach is desired.

I've also changed this to experimental because just so we get more examples before committing to an api.

cc: @Zayankovsky

Reply:
Ok changed this definition slightly. None of the tests needed to be changed, but I did add one to handle the new case

```
    @Test
    fun `onInitialize is called before first reference to the delegate`() {
        mockMapboxNavigationAppBehavior()
        val mockOnInitialize: (() -> Unit) = mockk(relaxed = true)
        val sut = SystemUnderTest(onInitialize = mockOnInitialize)

        MapboxNavigationApp.setup(mockk<NavigationOptions>())
        sut.moveToState(Lifecycle.State.CREATED)

        verify { mockOnInitialize.invoke() }
    }
```